overlays/nodejs/lambda/lambdaMixins.ts

```diff
@@ -0,0 +1,365 @@
+        closure, serializedFileNameNoExtension, options.codePathOptions);
+
+    // Copy over all option values into the function args.  Then overwrite anything we care
+    // about with our own values.  This ensures that clients can pass future supported
+    // lambda options without us having to know about it.
+    const copy = {
+        ...options,
+        code: new pulumi.asset.AssetArchive(codePaths),
+        handler: serializedFileNameNoExtension + "." + handlerName,
+        runtime: options.runtime || runtime.NodeJS8d10Runtime,
+        role: role.arn,
+        timeout: options.timeout === undefined ? 180 : options.timeout,
+    };
+
+    // Create the Lambda Function.
+    const lambda = new LambdaFunction(name, copy, opts);
+    lambda.roleInstance = role;
+    return lambda;
+}
+
+// computeCodePaths calculates an AssetMap of files to include in the Lambda package.
+async function computeCodePaths(
+        closure: Promise<pulumi.runtime.SerializedFunction>,
+        serializedFileNameNoExtension: string,
+        codePathOptions: pulumi.runtime.CodePathOptions | undefined): Promise<pulumi.asset.AssetMap> {
+
+    const serializedFunction = await closure;
+
+    // Construct the set of paths to include in the archive for upload.
+    let codePaths: pulumi.asset.AssetMap = {
+        // Always include the serialized function.
+        [serializedFileNameNoExtension + ".js"]: new pulumi.asset.StringAsset(serializedFunction.text),
+    };
+
+    // AWS Lambda always provides `aws-sdk`, so skip this.  Do this before processing user-provided
+    // extraIncludePackages so that users can force aws-sdk to be included (if they need a specific
+    // version).
+    codePathOptions = codePathOptions || {};
+    codePathOptions.extraExcludePackages = codePathOptions.extraExcludePackages || [];
+    codePathOptions.extraExcludePackages.push("aws-sdk");
+
+    const modulePaths = await pulumi.runtime.computeCodePaths(codePathOptions);
+
+    for (const [path, asset] of modulePaths) {
+        codePaths[path] = asset;
+    }
+
+    return codePaths;
+}
+
+const lambdaRolePolicy = {
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Action": "sts:AssumeRole",
+            "Principal": {
+                "Service": "lambda.amazonaws.com",
+            },
+            "Effect": "Allow",
+            "Sid": "",
+        },
+    ],
+};
+
+// sha1hash returns a partial SHA1 hash of the input string.
+function sha1hash(s: string): string {
+    const shasum: crypto.Hash = crypto.createHash("sha1");
+    shasum.update(s);
+    return shasum.digest("hex").substring(0, 8);
+}
+
+// Mixin the Role we potentially create into the Function instances we return.
+
+declare module "./function" {
+    interface Function {
+        // Actual Role instance value for this Function.  Will only be set if this function
+        // was created from [createFunction]
+        roleInstance?: iam.Role;
+    }
```

Comment:
This feels a little awkward - is this another reason to consider making this a subclass of `aws.lambda.Function`?

Reply:
Sure.  That seems very doable.